pkg/ingester/wal.go

```diff
@@ -774,39 +868,70 @@ Loop:
 		// With O(300 * #cores) in-flight sample batches, large scrapes could otherwise
 		// cause thousands of very large in flight buffers occupying large amounts
 		// of unused memory.
-		for len(record.Samples) > 0 {
+		for len(record.Samples) > 0 || len(walRecord.Samples) > 0 {
 			m := 5000
-			if len(record.Samples) < m {
-				m = len(record.Samples)
+			var userID string
+			if len(record.Samples) > 0 {
+				userID = record.UserId
+				if len(record.Samples) < m {
+					m = len(record.Samples)
+				}
+			}
+			if len(walRecord.Samples) > 0 {
+				userID = walRecord.UserID
+				if len(walRecord.Samples) < m {
+					m = len(walRecord.Samples)
+				}
 			}
+
 			for i := 0; i < params.numWorkers; i++ {
 				if len(shards[i].samples) == 0 {
 					// It is possible that the previous iteration did not put
 					// anything in this shard. In that case no need to get a new buffer.
-					shards[i].userID = record.UserId
+					shards[i].userID = userID
 					continue
 				}
 				select {
 				case buf := <-outputs[i]:
 					buf.samples = buf.samples[:0]
-					buf.userID = record.UserId
+					buf.userID = userID
 					shards[i] = buf
 				default:
 					shards[i] = &samplesWithUserID{
-						userID: record.UserId,
+						userID: userID,
 					}
 				}
 			}
```

Comment:
This seems like a lot of work for what's basically reusing of `samples` slice. It seems to me that pool would work equally well.

Reply:
I copied this logic as it is from Prometheus TSDB, and would like to keep any improvements to another PR. One reason I could think is that having a channel per worker reduces contention, else effectively a common pool would act as a bottleneck.